Fix exit in at_exit handlers

bew opened the pull request:
RFC!

There's no specs for `at_exit` stuff. Currently I made some basic test files to test the new behavior, I'll put them here in the description, along with their output, so we can discuss this new behavior, and if it's a good idea (or not).
I'll add them as specs later if needed.


#### Test: exit in a `at_exit` handler after a successful process execution changes the process's final status code

```cr
at_exit do |status|
  puts "1st at_exit handler, status is #{status}"
end

at_exit do |status|
  puts "2nd at_exit handler, status is #{status}"
  puts "Exiting with 2..."
  exit 2
end

at_exit do |status|
  puts "3rd at_exit handler, status is #{status}"
end
```
Exits with status `2`, outputs:
```
3rd at_exit handler, status is 0
2nd at_exit handler, status is 0
Exiting with 2...
1st at_exit handler, status is 2
```


#### Test: exit in a `at_exit` handler after an explicit `exit` call changes the process's final status code

```cr
at_exit do |status|
  puts "1st at_exit handler, status is #{status}"
end

at_exit do |status|
  puts "2nd at_exit handler, status is #{status}"
  puts "Exiting with 2..."
  exit 2
end

at_exit do |status|
  puts "3rd at_exit handler, status is #{status}"
end

exit 1
```
Exits with status `2`, outputs:
```
3rd at_exit handler, status is 1
2nd at_exit handler, status is 1
Exiting with 2...
1st at_exit handler, status is 2
```


#### Test: raise in a `at_exit` handler changes the status code if it was not already > 1

```cr
at_exit do |status|
  puts "1st at_exit handler, status is #{status}"
end

at_exit do |status|
  puts "2nd at_exit handler, status is #{status}"
  raise "Raised from at_exit !!"
end

at_exit do |status|
  puts "3rd at_exit handler, status is #{status}"
end
```
Exits with status `1`, outputs:
```
3rd at_exit handler, status is 0
2nd at_exit handler, status is 0
Error running at_exit handler: Raised from at_exit !!
1st at_exit handler, status is 1
```

Fixes https://github.com/crystal-lang/crystal/pull/5409#issuecomment-352797755

larubujo commented on src/crystal/main.cr:
```diff
@@ -56,7 +56,9 @@ module Crystal
 
-    AtExitHandlers.run status
+    new_status = AtExitHandlers.run status
+    status = {new_status, status}.max
```
last status should win, dont think max makes sense. ruby not work this way.

bew replied:
Yeah I'm not sure about this too, was just a try anyway!
My idea was: if the handler has an `exit` (or `exit 0`) and the process is exiting with a `exit 1` (or more), then it should not exit with a status code `0`.

No idea how Ruby works about that.

larubujo replied:
ruby seem to use status code from exit that was last executed. each exit changes status code

RX14 replied:
Indeed, please use the latest exist code.

straight-shoota commented:
Great, this prevents the error from #5409 wrongly reporting as exit status 0 in the circle CI build. So failed CI is actually confirming this change. =)

RX14 commented on src/kernel.cr:
```diff
@@ -160,12 +165,24 @@ def at_exit(&handler : Int32 ->) : Nil
 def exit(status = 0) : NoReturn
-  AtExitHandlers.run status
+  if AtExitHandlers.running?
+    raise NestedExitError.new status
```
it shouldn't be possible to rescue a nested exit, and it shouldn't use stack unwinding. Please create a state machine where calling `exit` simply resumes where the previous exit call left off.

RX14 commented on src/crystal/main.cr:
```diff
@@ -56,7 +56,9 @@ module Crystal
       end
 
-    AtExitHandlers.run status
+    new_status = AtExitHandlers.run status
```
Surely there would be a lot less duplication if this main function simply called `exit` with the `status` and never returned.

bew replied:
Good idea, I'll do that

bew replied:
But this will make `main` never return, that's probably not what we want IMO, it'll probably break some other things if we do that.

@ysbaddaden do you know something about this?
I'd like to have input from @asterite too (let's hope :smiley:)

bew commented:
Comment https://github.com/crystal-lang/crystal/pull/5413#discussion_r157902527 extracted:
> Context: main should just call `exit status` to reduce code duplication

But this will make `main` never return, that's probably not what we want IMO, it'll probably break some other things if we do that.

@ysbaddaden do you know something about this?
I'd like to have input from @asterite too (let's hope :smiley:)

RX14 commented:
@bew why would that break anything?

RX14 commented:
The only difference is between returning from `main` and calling `LibC.exit`. And libc does that anyway (yes, returning from `main` just means `_libc_main` will call `exit()` for you)

bew commented:
Hmm probably. I was thinking that it may break programs that redefine main, but I haven't put much thought about it actually.

RX14 commented:
Hmm, actually it probably will break those programs. My bad, I didn't think of that usecase.

Perhaps we should have a `:nodoc:` function shared between the two methods. Or perhaps we could define all the other items as `at_exit` handlers so `exit()` simply called `AtExitHandlers` then `LibC.exit`.

bew commented:
> we could define all the other items as at_exit handlers so exit() simply called AtExitHandlers then LibC.exit

That's also something I was thinking, done!

bcardiff commented on src/crystal/main.cr:
```diff
@@ -47,6 +47,11 @@ module Crystal
     GC.init
 
     remember_blocking_state
+    at_exit do
```
I am not totally convinced that handling the flush and restore of blocking state should be done with `at_exit`. I would rather leave some explicit calls for this. at_exit can't be cleared now, but might in the future. That way, also `@@handlers` would remain as `nil` in most of the cases.

bew replied:
Maybe we should have multiple kind of at exit handlers? Like the app's at exit handlers with the user's code (which can be cleared), and the crystal's at exit handlers?

bew replied:
nvm, it's just easier/faster to just duplicate the flushing+restore blocking, it's really simple code.

(was thinking of putting this in a class method in `Crystal` namespace (in `src/crystal/main.cr`) but I can't find a good name for it..)

bew commented:
@bcardiff I rebased, and reverted the use of `at_exit` to handle internal blocking state recovery.

RX14 commented on src/kernel.cr:
```diff
@@ -123,24 +123,28 @@ end
+    return status unless handlers = @@handlers
 
-    @@handlers.try &.reverse_each do |handler|
+    handler_iterator = @@handler_iterator ||= handlers.reverse_each
```
Actually, please write a note to explain we assign to `@@handler_iterator` to make sure we only call each handler once, even if an AtExitHandler calls `exit()`.

bew replied:
Good idea, done!

bew commented:
@RX14 I added the note about `@@handler_iterator` you requested, should be GTG!

RX14 commented:
Actually I think we already have specs which invoke the crystal compiler through `bin/crystal`.

bew commented:
I added some specs for `exit` & `at_exit`, using `bin/crystal` for now to compile the necessary snippets. Let me know what you think!

bew commented:
Can someone restart the Travis build? It failed due to a `fork: Cannot allocate memory` error, on an unrelated spec.

bew commented:
@asterite why is it important to print the exception **after** the exit handlers?

asterite commented:
One reason is that if you have an error in your program and you have an exit handler print stuff, the error might be hidden after all those messages.

Another way to think about this is that printing the error is the first exit handler, and other exit handlers are added next. Then, exit handlers run in reverse order, so printing the exception should come last.

The final reason is that Ruby does it this way, so there must be a good reason.

RX14 commented on src/kernel.cr:
```diff
@@ -123,24 +123,31 @@ end
 module AtExitHandlers
-  @@running = false
-
   def self.add(handler)
```
This should probably raise while we're iterating (i.e. `@@handler_iterator != nil`) to stop people modifying the exit handlers in an exit handler.

asterite replied:
I just tried it in Ruby, it works. We just need to do a loop with an index (in case it doesn't work already).

asterite replied:
I mean, this works in Ruby:

```ruby
at_exit do
  at_exit do
    puts "BYE"
  end
  puts "EXIT"
end
```

Ouptut:

```
EXIT
BYE
```

RX14 replied:
I don't think it should work. It's confusing and creates for actively confusing code. Let's not follow ruby's example on this.

bew replied:
What would be the usecase of having at exit handler adding new at exit handlers?

asterite replied:
I see nothing wrong with it working. It's intuitive and super easy to implement. But if you think it should be prohibited, it's fine too (but I see nothing wrong with it).

bew replied:
For the record, in ruby the nested at_exit handlers are run just after the handler around them:
```ruby
at_exit do
  puts "first"
end

at_exit do
  at_exit do
    puts "> nested"
  end
  puts "second"
end

at_exit do
  puts "third"
end
```
Gives:
```
third
second
> nested
first
```
---
Not sure it can be useful for anything with this behavior...

asterite replied:
Yes, `at_exit` adds a handler to the end of the list, and to traverse them we should use `pop?`. Super simple. Right now it uses `reverse_each`, but a change in my local copy to use `pop?` made it work.

RX14 replied:
I agree that this is how it should work if we want this to work.

However I don't see the usecase, and so I'd rather be more restrictive until we knew what the usecases were. This is because it creates a far more complex and dynamic control flow through the handlers which may be hard to track and hard to debug. So i'd err on the side of saying "no".

asterite replied:
Sounds good. We can wait until someone complaints with a real use case.

RX14 replied:
That being said, the `pop?` code is robust whether we disallow this or not, and is smaller and easier to read. If we decide to enable it it should be as easy as removing a `raise ... if` line of code.

bew commented:
I see.. One issue which can arise is in the case where an exit handler calls `exit`, the execution won't come back to `main` _and the exception might never be printed_.

I think that a way to ensure the exception is properly printed even in case of `exit`, would be to insert an exit handler (at position 0) that would print the exception (like the thinking mode suggestion from @asterite).

bew commented:
I refactored the handler loop with `pop?`, I agree, it's simpler and better! thanks @asterite 

I also disabled nested at_exit handlers and added a spec for it @RX14.

bew commented:
Ping :smiley: 

@asterite I'd like to have your opinion about my previous comment https://github.com/crystal-lang/crystal/pull/5413#issuecomment-355277715 thank you!

bew commented:
I rebased the commits.

And changed the exception printing to be after all other at_exit handlers @asterite :smiley:

Sija commented on src/crystal/main.cr:
```diff
@@ -56,11 +56,15 @@ module Crystal
+    # This will run after all at_exit handlers, to make sure the user sees the exception.
+    AtExitHandlers.do_last ->(code : Int32) do
+      exception = ex
+      exception.inspect_with_backtrace STDERR if exception
```
`ex.try(&.inspect_with_backtrace(STDERR)`?

bew replied:
Yes! that's a good idea, thank you!

bew commented:
And @bcardiff `test_darwin` is broken with a weird error:
```
  1) Process returns status 127 if command could not be executed
     Failure/Error: process.wait.exit_code.should eq(127)

       Expected: 127
            got: 0

     # spec/std/process_spec.cr:18
```
It probably comes from my code, but I don't see what could be the difference that change this specific behavior on macos..

bew commented:
Ok, I managed to test this on a macbook, and pushed a commit that mysteriously fixed the spec..

Let's wait for CircleCI, but if it really fixed it, I would like to know why if someone knows..

RX14 commented:
I'd like to merge this once you rebase on top of #5804

bew commented:
Rebased on master!

As per [this comment](https://github.com/crystal-lang/crystal/pull/5413#discussion_r158746689) from @bcardiff:
> at_exit can't be cleared now, but might in the future

To make sure the exception's handler don't get clear if this becomes true, I used a separate list to store the exception(s).

And added a spec for an unhandled exception.

RX14 commented:
CI failed :(

bew commented:
Nice, it got a tricky case that I missed!

Sija commented on src/kernel.cr:
```diff
@@ -125,22 +125,47 @@ end
+    if handlers = @@handlers
+      # Run the registered handlers in reverse order
+      while handler = handlers.pop?
+        begin
```
IIRC you can skip `begin ... end` block altogether:

```cr
while handler = handlers.pop?
  # ...
rescue handler_ex
  # ...
end
```

bew replied:
I thought I worked too, but it doesn't inside a while..

bew replied:
I don't remember exactly but I think it was ambiguous whether the rescue should be after each loop, or after the whole while.

straight-shoota replied:
yep

bew commented:
I have this waiting in my fork, should I include it in this PR?
https://github.com/bew/crystal/compare/fix-exit-in-at_exit-handlers...print-full-stacktrace-of-at_exit-errors

It just print the full stacktrace when an exception occurs in a at_exit handler (instead of just the message)

bew commented:
Anything left to do?

RX14 commented on src/kernel.cr:
```diff
@@ -125,22 +125,47 @@ end
   @@running = false
 
+  private class_getter(handlers) { [] of Int32 -> }
+  private class_getter(exceptions) { [] of Exception }
```
When would you have more than one?

bew replied:
Currently never, I was just thinking that maybe some shards might want to add multiple exceptions.. But now that I rethink about it it doesn't make sense, or at least I shoukd apply YAGNI until someone needs this.

RX14 commented on src/kernel.cr:
```diff
@@ -125,22 +125,47 @@ end
+
   def self.add(handler)
-    handlers = @@handlers ||= [] of Int32 ->
+    raise "Cannot use at_exit from an at_exit handler" if @@running
```
"in an" not "from an"

bew commented on src/kernel.cr:
```diff
@@ -171,10 +196,9 @@ end
+  status = AtExitHandlers.run status
   STDOUT.flush
   STDERR.flush
-  Crystal.restore_blocking_state
```
Note for myself: don't forget to re-add this line

bew commented:
I've rebased, and applied @RX14's comments, should be GTG I think when CI passes!

bew commented:
Cool thank you!
